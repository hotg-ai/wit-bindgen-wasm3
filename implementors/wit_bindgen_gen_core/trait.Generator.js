(function() {var implementors = {};
implementors["wit_bindgen_gen_wasm3"] = [{"text":"impl <a class=\"trait\" href=\"wit_bindgen_gen_core/trait.Generator.html\" title=\"trait wit_bindgen_gen_core::Generator\">Generator</a> for <a class=\"struct\" href=\"wit_bindgen_gen_wasm3/struct.Wasmtime.html\" title=\"struct wit_bindgen_gen_wasm3::Wasmtime\">Wasmtime</a>","synthetic":false,"types":["wit_bindgen_gen_wasm3::Wasmtime"]}];
if (window.register_implementors) {window.register_implementors(implementors);} else {window.pending_implementors = implementors;}})()