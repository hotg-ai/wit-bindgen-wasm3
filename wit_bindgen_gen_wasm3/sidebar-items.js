initSidebarItems({"enum":[["Async",""]],"struct":[["Opts",""],["Wasmtime",""]]});