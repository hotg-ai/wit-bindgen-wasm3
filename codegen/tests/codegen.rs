#[macro_use]
#[cfg(test)]
extern crate pretty_assertions;

use std::{collections::HashMap, path::Path};

use wit_bindgen_gen_core::{wit_parser::Interface, Files, Generator};
use wit_bindgen_gen_wasm3::Opts;

#[test]
fn single_function() {
    let path = Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("tests")
        .join("data")
        .join("single-function.wit");
    let single_function = Interface::parse_file(&path).unwrap();
    let mut files = Files::default();

    Opts::default()
        .build()
        .generate_all(&[single_function], &[], &mut files);

    let files: HashMap<_, _> = files.iter().collect();
    assert_eq!(files.len(), 1);
    let generated = std::str::from_utf8(files["bindings.rs"]).unwrap();
    println!("=========\n{}=======", generated);
    assert_eq!(generated, "asd");
}
