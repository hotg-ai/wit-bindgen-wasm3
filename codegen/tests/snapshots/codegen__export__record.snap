---
source: codegen/tests/codegen.rs
expression: record
---
pub mod record {
    #[allow(unused_imports)]
    use wit_bindgen_wasm3::rt::{anyhow, wasm3};
    #[derive(Clone)]
    pub struct Foo<'a> {
        pub bar: &'a str,
    }
    impl<'a> std::fmt::Debug for Foo<'a> {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            f.debug_struct("Foo").field("bar", &self.bar).finish()
        }
    }
    pub trait Record: Sized {
        fn start(&mut self, value: Foo<'_>);
    }

    pub fn register<U>(
        module: &wit_bindgen_wasm3::rt::wasm3::Module<'_>,
        state: U,
    ) -> wit_bindgen_wasm3::rt::wasm3::error::Result<()>
    where
        U: Record + Send + Sync + 'static,
    {
        use wit_bindgen_wasm3::rt::get_memory;
        let state = std::sync::Arc::new(std::sync::Mutex::new(state));
        {
            let state = std::sync::Arc::clone(&state);
            module.link_closure(
                "record",
                "start",
                move |mut caller: wit_bindgen_wasm3::rt::wasm3::CallContext,
                      (arg0, arg1): (i32, i32)| {
                    let memory = unsafe { &*caller.memory_mut() };
                    let mut _bc = wit_bindgen_wasm3::BorrowChecker::new(memory);
                    let state = state.lock().expect("lock was poisoned");
                    let ptr0 = arg0;
                    let len0 = arg1;
                    let param0 = Foo {
                        bar: _bc.slice_str(ptr0, len0)?,
                    };
                    state.start(param0);
                    Ok(())
                },
            )?;
        }
        Ok(())
    }
}

